_posts/2021-11-19-migrating-kafka-with-mirror-maker2.md

```diff
@@ -0,0 +1,227 @@
+    migrated **must** be aligned to the Kafka API declared, otherwise
+    Apache Kafka brokers will have to perform down-conversion
+    continuously (not desirable as it will impact in the performance)
+4. Deploy MirrorMaker in the target OpenShift platform.
+5. Verify topics and data are created in the target Apache Kafka cluster.
+6. Move consumers from the source Apache Kafka cluster to the target
+Apache Kafka cluster. (Remember Consumer First approach)
+7. Verify data consumption is working successfully.
+8. Move producers from the source Apache Kafka cluster to the target
+Apache Kafka cluster.
+9. Verify data production is working successfully.
+
+The migration process could add many different required or optional
+tasks aligned with the details or specific activities, but in general
+the process will be very similar to this one.
+
+Now, let’s see how we could execute it.
+
+## MirrorMaker, the migration tool
+
+The Apache Kafka ecosystem provides a tool to replicate data between
+different Apache Kafka clusters: [Apache MirrorMaker](https://kafka.apache.org/documentation/#georeplication).
+
+**NOTE**: There are other approaches to migrate Apache Kafka workloads
+based on other tools and processes, but all of them are out of the
+scope. We think that the Apache Kafka ecosystem provides the right tools
+to help in this scenario.
+
+This tool helps to define data flows that cross the boundaries of an
+Apache Kafka cluster to extend to others, allowing to cover different uses
+cases and scenarios such as:
+
+* Geo-replication
+* Disaster recovery
+* Feeding edge cluster into a central, aggregate cluster
+* Cloud migration or hybrid cloud deployments (we are here)
+* Legal, technical, or compliance requirements
+
+MirrorMaker provides currently two different implementations:
+
+* **MirrorMaker 1**: The first implementation with basic functionality
+to consume and produce data from the source to the target. It is very
+limited and it could help you to replicate only data payloads. 
+* **MirrorMaker 2**: Improved implementation based in
+[KafkaConnect framework](https://kafka.apache.org/documentation/#connect)
+allowing  replication of more things (topics and its configuration,
+consumer groups offsets, ACLs), and more features (detects new topics/partitions,
+metrics, fault-tolerant and HA).
+
+This article is based on MirrorMaker2 because it will include a set of
+features needed to cover our requirements. The most important features are:
+
+* **Topic topology replication**, as we need to have the same structure of
+the topic and partitions in the target Kafka cluster.
+* **Consumer group offset synchronization**, as we need to consume the data
+from the target Kafka cluster at the same point the consumers stopped in
+the source Kafka cluster. This feature tries to avoid consuming duplicated
+messages in the target cluster.
+* **Dynamic detection of changes** in source Kafka cluster and replicate in
```

Comment:
we are talking about detection of configuration changes on brokers right? should be clearer here?